src/refs.rs

```diff
@@ -0,0 +1,408 @@
+        let expected = [
+            (root, dag0),
+            (dag0, unixfs0),
+            (dag0, dag1),
+            (dag1, unixfs1),
+            (root, unixfs0),
+            (root, dag1),
+            (dag1, unixfs1),
+            (root, unixfs1),
+        ];
+
+        println!("found edges:\n{:#?}", all_edges);
+
+        assert_edges(&expected, all_edges.as_slice());
+    }
+
+    #[tokio::test(max_threads = 1)]
+    async fn all_unique_refs_from_root() {
+        let Node { ipfs, bg_task: _bt } = preloaded_testing_ipfs().await;
+
+        let (root, dag0, unixfs0, dag1, unixfs1) = (
+            // this is the dag with content: [dag0, unixfs0, dag1, unixfs1]
+            "bafyreihpc3vupfos5yqnlakgpjxtyx3smkg26ft7e2jnqf3qkyhromhb64",
+            // {foo: dag1, bar: unixfs0}
+            "bafyreidquig3arts3bmee53rutt463hdyu6ff4zeas2etf2h2oh4dfms44",
+            "QmPJ4A6Su27ABvvduX78x2qdWMzkdAYxqeH5TVrHeo3xyy",
+            // {foo: unixfs1}
+            "bafyreibvjvcv745gig4mvqs4hctx4zfkono4rjejm2ta6gtyzkqxfjeily",
+            "QmRgutAxd8t7oGkSm4wmeuByG6M51wcTso6cubDdQtuEfL",
+        );
+
+        let root_block = ipfs.get_block(&Cid::try_from(root).unwrap()).await.unwrap();
+        let ipld = decode_ipld(root_block.cid(), root_block.data()).unwrap();
+
+        let destinations: HashSet<_> = iplds_refs(ipfs, vec![(root_block.cid, ipld)], None, true)
+            .map_ok(|Edge { destination, .. }| destination.to_string())
+            .try_collect()
+            .await
+            .unwrap();
+
+        // go-ipfs output:
+        // bafyreihpc3vupfos5yqnlakgpjxtyx3smkg26ft7e2jnqf3qkyhromhb64 -> bafyreidquig3arts3bmee53rutt463hdyu6ff4zeas2etf2h2oh4dfms44
+        // bafyreihpc3vupfos5yqnlakgpjxtyx3smkg26ft7e2jnqf3qkyhromhb64 -> QmPJ4A6Su27ABvvduX78x2qdWMzkdAYxqeH5TVrHeo3xyy
+        // bafyreihpc3vupfos5yqnlakgpjxtyx3smkg26ft7e2jnqf3qkyhromhb64 -> bafyreibvjvcv745gig4mvqs4hctx4zfkono4rjejm2ta6gtyzkqxfjeily
+        // bafyreihpc3vupfos5yqnlakgpjxtyx3smkg26ft7e2jnqf3qkyhromhb64 -> QmRgutAxd8t7oGkSm4wmeuByG6M51wcTso6cubDdQtuEfL
+
+        let expected = [dag0, unixfs0, dag1, unixfs1]
+            .iter()
+            .map(|&s| String::from(s))
+            .collect::<HashSet<_>>();
+
+        let diff = destinations
+            .symmetric_difference(&expected)
+            .map(|s| s.as_str())
+            .collect::<Vec<&str>>();
+
+        assert!(diff.is_empty(), "{:?}", diff);
+    }
+
+    fn assert_edges(expected: &[(&str, &str)], actual: &[(String, String)]) {
+        let expected: HashSet<_> = expected.iter().map(|&(a, b)| (a, b)).collect();
+
+        let actual: HashSet<_> = actual
+            .iter()
+            .map(|(a, b)| (a.as_str(), b.as_str()))
+            .collect();
+
+        let diff: Vec<_> = expected.symmetric_difference(&actual).collect();
+
+        assert!(diff.is_empty(), "{:#?}", diff);
+    }
+
+    async fn preloaded_testing_ipfs() -> Node {
```

Comment:
this becomes duplicated in http/src/v0/refs.rs and here, but I figured it's better to leave duplicated since the tests over at http aren't duplicating what functionality is here.

the http tests should be however against the warp handlers but only the refs_local is being directly tested.